File: mbox_project_remind/ex28_bin/phase_1/ex28_bin_2.py
import sys
import struct

L_WIDTH = 25
R_WIDTH = 25

"""
要件：
- `struct`モジュールを使用すること
- コマンドライン引数でファイル名を受け取ること
- ファイル識別子が"BM"でない場合はエラーメッセージを表示すること
- 各情報にラベルを付けて見やすく整形すること
- ファイルサイズはバイト単位に加えて、適切な単位(KB, MB)でも表示すること


期待される出力：

=== BMP Header Information ===
File Type: BM
File Size: 30054 bytes (29.3 KB)
Image Offset: 54 bytes
Image Width: 100 pixels
Image Height: 100 pixels
Bit Depth: 24 bits
Compression: None (0)

"""

def print_result(title, label, value):
    if title is not None:
        title = title.center(30, '=')
        print(title)
    print(f"{label + ':':<{L_WIDTH}}{value:>{R_WIDTH}}")





def main():
    if len(sys.argv) != 2:
        print("Argument Error", file=sys.stderr)
        print("Usage: [This file] [BMP file]", file=sys.stderr)
        return 1

    file_name = sys.argv[1]

    try:
        with open(file_name, 'rb') as f:
            f.seek(0) # ファイルのアドレスを指定のbyteに合わせる関数？
            file_header = f.read(14)

            # char*2, 符号なし4byte, 4byteスキップ、符号なし4byte
            signature, file_size, img_offset = struct.unpack('<2sIxxxxI', file_header)

            print(signature, file_size, img_offset, file=sys.stderr)

            if signature != b'BM':
                print('File is not BMP', file=sys.stderr)
                return 1

            # 画像サイズまで読む
            # 終端idx - 開始idx は距離が求められるという事を理解する
            f.seek(18)
            file_detail = f.read(8)
            img_width, img_hight = struct.unpack('<II', file_detail)

            print(img_width, img_hight, file=sys.stderr)

            # 圧縮情報まで読む
            f.seek(28)
            file_detail = f.read(6)
            bit_depth, compression = struct.unpack('<HI', file_detail)

            print(bit_depth, compression, file=sys.stderr)

            # 標準出力
            title = ' BMP Header Information '
            print_result(title, 'File Type' , str(signature))

            file_size_kb = f"{int(file_size) / 1024:.1f}"
            print_result(None, 'File Size', f"{str(file_size) + ' bytes (' + str(file_size_kb) + ' KB)'}")
            print_result(None, 'Imagine Offset', f"{str(img_offset) + ' bytes'}")
            print_result(None, 'Imagine Width', f"{str(img_width) + ' pixels'}")
            print_result(None, 'Imagine Width', f"{str(img_hight) + ' pixels'}")
            print_result(None, 'Bit Depth', f"{str(bit_depth) + ' bits'}")
            print_result(None, 'Compression', f"{str(compression)}")
            print()

            return 0

    except FileNotFoundError as e:
        print(e, file=sys.stderr)
        return 1
    except Exception as e:
        print(e, file=sys.stderr)
        return 1

if __name__ == '__main__':
    sys.exit(main())

File: mbox_project_remind/ex28_bin_phase_1/ex28_bin_2.c
#include <stdio.h>
#include <stdint.h>

// データをパディングせず「1byteごとに詰めて保存する」
// また、既存の設定を保存(push)しておく
#pragma pack(push, 1)

typedef struct {
	uint16_t	file_type;
	uint32_t	file_size;
	uint16_t	reserved_1;
	uint16_t	reserved_2;
	uint32_t	pixel_offset;
} BmpFileHeader;

typedef struct {
	uint32_t	header_size;
	int32_t		pixel_width;	// bmpの幅・高さは符号付きで保存する
	int32_t		pixel_hight;
	uint16_t	planes;
	uint16_t	bit_depth;
	uint32_t	compression;
	// 後は使わないので省略
} BmpInfoHeader;

// あらかじめ保存(push)していた既存設定に戻す
// LIFO: Last in First Out 後入れ先出し
#pragma pack(pop)

/*
- ファイルサイズ
- 画像の幅と高さ
- 1ピクセルあたりのビット数
- 圧縮方式
- ピクセルデータへのオフセット
*/

int	main(int argc, char **argv) {

	if (argc != 2) {
		fprintf(stderr, "Argument error\n");
		fprintf(stderr, "Usage: [this file] [.bmp]\n");
		return 1;
	}

	const char	*file_name = argv[1];
	FILE		*fp = fopen(file_name, "rb");
	if (fp == NULL) {
		fprintf(stderr, "Cannot open file\n");
		return 1;
	}

	BmpFileHeader	file_header;
	// バッファする構造体、構造体のサイズ(読み込み量)、バッファの数、読み込むファイル
	// 成功したら、引数のバッファ数を返す
	if (fread(&file_header, sizeof(file_header), 1, fp) != 1) {
		fprintf(stderr, "Cannot read file_header\n");
		fclose(fp);
		return 1;
	}

	// 0x4d42 == BM .bmpかどうか確認
	if (file_header.file_type != 0x4d42) {
		fprintf(stderr, "File is not .bmp\n");
		fclose(fp);
		return 1;
	}

	BmpInfoHeader	info_header;
	if (fread(&info_header, sizeof(info_header), 1, fp) != 1) {
		fprintf(stderr, "Cannot read info_header\n");
		fclose(fp);
		return 1;
	}

	// プレーン数は必ず１なので、それで整合性を確認
	if (info_header.planes != 0x01) { // 16進数で確認！！
		fprintf(stderr, "Plane number is not 1\n");
		fclose(fp);
		return 1;
	}

	/*
	=== BMP File Information ===
	File size: 310 bytes
	Image size: 8 x 8 pixels
	Dimensions: 8 x 8 pixels
	Bits per pixel: 24
	Compression: 0 (0=none)
	Data offset: 54 bytes
	*/

	printf("=== BMP File Information ===\n");
	printf("File size: %d bytes\n", file_header.file_size);
	printf("Image size: %d * %d pixels\n", info_header.pixel_width, info_header.pixel_hight);
	printf("Dimensions: %d * %d pixels\n", info_header.pixel_width, info_header.pixel_hight);
	printf("Bit per pixels: %d\n", info_header.bit_depth);
	if (info_header.compression == 0x00)
		printf("Compression: 0 (0=None)\n");
	else
		printf("Compression: %d\n", info_header.compression);
	printf("Data offset: %d bytes\n", file_header.pixel_offset);
	printf("\n");

	fclose(fp);
	return 0;
}
